index.bs

```diff
@@ -854,6 +884,12 @@ Note: The {{XRSession}}'s [=visibility state=] does not necessarily imply the vi
 
 Note: The {{XRSession}}'s [=visibility state=] does not affect or restrict mouse behavior on tethered sessions where 2D content is still visible while an [=immersive session=] is active. Content should consider using the [[!pointerlock]] API if it wishes to have stronger control over mouse behavior.
 
+Each {{XRSession}} MAY have an internal <dfn for="XRSession">frameRate</dfn> which is the rate per second at which the {{XRSession}} SHOULD produce frames. If the user agent is unable to produce frames each second for this rate, the [=XR Compositor=] MAY use reprojection or other techniques to improve the experience. This attribute is optional and MUST NOT be present for {{XRSessionMode/inline}} sessions. If [=XRSession/frameRate=] is <code>0</code>, the user agent is allowed to dynamically change the frame rate.
+
+The <dfn attribute for="XRSession">targetFrameRate</dfn> attribute reflects the internal [=XRSession/framerate=]. If the {{XRSession}} has no [=XRSession/framerate=], return `null`. The <dfn attribute for="XRSession">ontargetframeratechange</dfn> attribute is an [=Event handler IDL attribute=] for the {{targetframeratechange}} event type. If the {{XRSession}}'s frame rate is updated for any reason, it MUST [=apply the target frame rate=] with the new frame rate.
+
+The <dfn attribute for="XRSession">supportedFrameRates</dfn> attribute returns a list of supported frame rates. This attribute is optional and MUST NOT be present for {{XRSessionMode/inline}} sessions. If the user agent can dynamically change the frame rate, this should be reflected as a frame rate of <code>0</code>. If the {{XRSession}} has an internal [=XRSession/framerate=], it also MUST support the {{XRSession/supportedFrameRates}} attribute.
```

Comment:
> If the {{XRSession}}'s frame rate is updated for any reason, it MUST [=apply the target frame rate=] with the new frame rate.

I don't fully follow when this applies.

For example, on HoloLens, the app cannot affect the frame rate in any way - in normal usage, the app's render rate will always be 60Hz.  If the app misses frames, previous frames can be reprojected, but the app continues to be expected to keep up.  Specifically, I would expect `supportedFrameRates` to only contain `60`, because the app cannot request any other frame timings.

However, if the user starts a Mixed Reality Capture video recording, the system will drop to 30Hz all-up, and the app will start to receive 30Hz frame timing.  Is this sentence saying that because the effective frame rate got updated "for any reason", that an `ontargetframeratechange` event must be fired and `targetFrameRate` must now return `30`?  This seems wrong, since `30` is not an element within `supportedFrameRates`.

> If the user agent can dynamically change the frame rate, this should be reflected as a frame rate of <code>0</code>.

This clause sets the wrong mental model for developers.  This makes it sound like the target frame timing is fixed if the app chooses a non-`0` value, and dynamic if the app chooses `0`.  In reality, things are always somewhat dynamic regardless - the UA can make use of the `targetFrameRate` hint provided by the page, but any given frame can have a duration that either matches or doesn't match that target for various platform-specific reasons.

It seems more flexible and consistent to treat `targetFrameRate` as a hint to the UA, with the page able to observe for any given frame what the effective frame rate/duration is, which can change whether the page has set a non-`0` target or not.

Perhaps an answer here is to change the API to `requestTargetFrameRate`, which returns a promise that completes when the UA has taken action on the request to its best ability.  No guarantee is made that the frame rate won't change further for other reasons after the request (e.g. halved due to system overlay, background app, this app not hitting framerate, thermal condition, etc.).  However, if the app is waiting until the frame rate logically switches to 72Hz before starting video playback, the `requestTargetFrameRate` promise completion is the time to start.  We could then choose to additionally surface on `XRFrame` a given frame's duration, without promising that it will be an element of `supportedFrameRates`, although perhaps it is cleaner first to see if devs are satisfied with just the promise.

Reply:
> > If the {{XRSession}}'s frame rate is updated for any reason, it MUST [=apply the target frame rate=] with the new frame rate.
> 
> I don't fully follow when this applies.
> 
> For example, on HoloLens, the app cannot affect the frame rate in any way - in normal usage, the app's render rate will always be 60Hz. If the app misses frames, previous frames can be reprojected, but the app continues to be expected to keep up. Specifically, I would expect `supportedFrameRates` to only contain `60`, because the app cannot request any other frame timings.

`supportedFrameRates` is the list of framerates that the author can set. They are passed to the system and it will decide how (or if) to apply it. In the case of Hololens, this list could just `60` or it can decide not to support this function which means that the app can't control frame rate.

> However, if the user starts a Mixed Reality Capture video recording, the system will drop to 30Hz all-up, and the app will start to receive 30Hz frame timing. Is this sentence saying that because the effective frame rate got updated "for any reason", that an `ontargetframeratechange` event must be fired and `targetFrameRate` must now return `30`? This seems wrong, since `30` is not an element within `supportedFrameRates`.

Since the rate the author sets is just a hint for the compositor, it is free to change it.
Quest also has cases where it drops frame rates. For instance, if the user presses the Oculus button on the controller to bring up the menu, the rate is cut by half.

> > If the user agent can dynamically change the frame rate, this should be reflected as a frame rate of `0`.
> 
> This clause sets the wrong mental model for developers. This makes it sound like the target frame timing is fixed if the app chooses a non-`0` value, and dynamic if the app chooses `0`. In reality, things are always somewhat dynamic regardless - the UA can make use of the `targetFrameRate` hint provided by the page, but any given frame can have a duration that either matches or doesn't match that target for various platform-specific reasons.

I'm happy to remove this special case.

> It seems more flexible and consistent to treat `targetFrameRate` as a hint to the UA, with the page able to observe for any given frame what the effective frame rate/duration is, which can change whether the page has set a non-`0` target or not.

Yes, this is what I intended to do. Maybe I need it to make it more explicit and also add a note?

> Perhaps an answer here is to change the API to `requestTargetFrameRate`, which returns a promise that completes when the UA has taken action on the request to its best ability. No guarantee is made that the frame rate won't change further for other reasons after the request (e.g. halved due to system overlay, background app, this app not hitting framerate, thermal condition, etc.). However, if the app is waiting until the frame rate logically switches to 72Hz before starting video playback, the `requestTargetFrameRate` promise completion is the time to start. We could then choose to additionally surface on `XRFrame` a given frame's duration, without promising that it will be an element of `supportedFrameRates`, although perhaps it is cleaner first to see if devs are satisfied with just the promise.

I prefer to have an event instead of a promise to give the UA the flexibility to change the rate at any point, even without setting a target frame rate. 
For instance, on Hololens this event would fire if Mixed Reality Capture starts.

Reply:
> > > If the user agent can dynamically change the frame rate, this should be reflected as a frame rate of 0.
> >
> > This clause sets the wrong mental model for developers. This makes it sound like the target frame timing is fixed if the app chooses a non-0 value, and dynamic if the app chooses 0. In reality, things are always somewhat dynamic regardless - the UA can make use of the targetFrameRate hint provided by the page, but any given frame can have a duration that either matches or doesn't match that target for various platform-specific reasons.
>
> I'm happy to remove this special case.

Yea - I would eliminate ever enumerating `0` from `supportedFrameRates`.  We still need to define the default `null` state for `targetFrameRate` (and allow apps to set it again when video playback ends), though that can just mean that the app is expressing no preference to the UA.

> I prefer to have an event instead of a promise to give the UA the flexibility to change the rate at any point, even without setting a target frame rate.
For instance, on Hololens this event would fire if Mixed Reality Capture starts.

Got it - the name `ontargetframeratechange` suggested to me that the event could only return values from `supportedFrameRates`, and perhaps only triggered when the app had requested a change to the target frame rate.

Perhaps just call this `onframeratechange` to make clear that it's about the effective frame rate, not the target?

Reply:
> Got it - the name `ontargetframeratechange` suggested to me that the event could only return values from `supportedFrameRates`, and perhaps only triggered when the app had requested a change to the target frame rate.
> 
> Perhaps just call this `onframeratechange` to make clear that it's about the effective frame rate, not the target?

In the case of Oculus, the system still draws at the full frame rate. It's just the browser that is asked to slow down. Given that, I'm leaning toward `ontargetframeratechange` but I would be fine either way.